core/sdk-core/src/test/java/software/amazon/awssdk/core/document/VoidDocumentVisitorTest.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.document;
+
+import org.testng.annotations.Test;
+import software.amazon.awssdk.core.SdkNumber;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class VoidDocumentVisitorTest {
+    @Test
+    public void voidDocumentVisitor() {
+        VoidDocumentVisitCount voidDocumentVisitor = new VoidDocumentVisitCount();
+        // Constructing a Document for CustomClass
+        final Document build = getCustomDocument();
+        build.accept(voidDocumentVisitor);
+        assertThat(voidDocumentVisitor.booleanVisitsVisits).isEqualTo(2);
+        assertThat(voidDocumentVisitor.mapVisits).isEqualTo(2);
+        assertThat(voidDocumentVisitor.nullVisits).isEqualTo(1);
+        assertThat(voidDocumentVisitor.listVisits).isEqualTo(1);
+        assertThat(voidDocumentVisitor.numberVisits).isEqualTo(4);
+        assertThat(voidDocumentVisitor.stringVisits).isEqualTo(2);
+        // Expected  CustomClass
+    }
+
+    public Document getCustomDocument() {
+        final Document build = Document.mapBuilder()
+                .putDocument("customClassFromMap",
+                        Document.mapBuilder().putString("innerStringField", "innerValue")
+                                .putNumber("innerIntField", SdkNumber.fromLong(1)).build())
+                .putString("outerStringField", "outerValue")
+                .putNull("nullKey")
+                .putBoolean("boolOne", true)
+                .putBoolean("boolTwo", false)
+                .putList("listKey", listBuilder -> listBuilder.addNumber(SdkNumber.fromLong(2)).addNumber(SdkNumber.fromLong(3)))
+                .putNumber("outerLongField", SdkNumber.fromDouble(4)).build();
+        return build;
+    }
+
+
+    private static class VoidDocumentVisitCount implements VoidDocumentVisitor {
```

Comment:
minor: in the future, it might be easier to use Mockito to count the invocations instead https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#4

Reply:
Yeah sure.I initially tried using mock but realized it was taking more time, since the interfaces were of void type I was not able to directly use doNothing() on them because for List and Map we need to visit individual documents , which makes visitor to call  doCallRealMethod. 
Will change this once we have support for ```doCallRealMethod().when(instance).voidFunction();```